### ISSUE TITLE
`NaN` padding in zarr files with delayed start times 

### ISSUE THREAD

**Issue by nicrie** [OceanParcels/parcels#1487]
Following up on the [discussion](https://github.com/OceanParcels/parcels/discussions/1485) regarding potential issues with suboptimal chunking in zarr files generated by `parcels`. When setting a chunk size larger than the total number of released particles (or observations), the resulting zarr file is padded with `NaN` values for trajectories/observations. This results in a file shape of `(n_chunks_particles + n_particles_per_day, n_chunks_observations + 1)`.

**Note**: this only happens when setting delayed starting times. If starting time is the same for all released particles, the shape of the output file is correct.

### Minimal Working Example

For clarity, I've simplified my example using the `parcels` tutorial data. Here, 100 particles are released daily over a rectangular area for 5 days, totaling 500 particles and 5 observations.

When the chunk size is set smaller than the total number of released particles/observations, the dimensions of the output zarr file match expectations: `Frozen({'trajectory': 500, 'obs': 5})`.

However, for larger chunk sizes (e.g., `chunks_particles=10_000` and `chunks_obs=100`), the output zarr file dimensions are: `Frozen({'trajectory': 10,100, 'obs': 101})`.

By removing all NaN trajectories and observations using `ds.dropna("obs", "all").dropna("trajectory", "all").dims`, the dimensions revert to the expected values: `Frozen({'trajectory': 500, 'obs': 5})`.

<details>

<summary>Code </summary>

```py
import os
from datetime import datetime, timedelta

import numpy as np
import xarray as xr

from parcels import (
    AdvectionRK4,
    FieldSet,
    ParcelsRandom,
    ParticleSet,
    JITParticle,
    download_example_dataset,
)

# Define fieldset
example_dataset_folder = download_example_dataset("GlobCurrent_example_data")
fname = f"{example_dataset_folder}/*.nc"
filenames = {"U": fname, "V": fname}
variables = {
    "U": "eastward_eulerian_current_velocity",
    "V": "northward_eulerian_current_velocity",
}
dimensions = {
    "U": {"lat": "lat", "lon": "lon", "time": "time"},
    "V": {"lat": "lat", "lon": "lon", "time": "time"},
}

fieldset = FieldSet.from_netcdf(
    filenames, variables, dimensions, deferred_load=True, chunksize=False
)


# Number of particles released per day
n_particles_per_day = 100

# Chunking
chunks_particles = 10_000
chunks_obs = 100


# Target directory
dir_target = "parcels/trajs/"
if not os.path.exists(dir_target):
    os.makedirs(dir_target)

# Release and runtime
runtime_total = timedelta(days=5)
runtime_dt = timedelta(days=1).total_seconds()
outputdt = timedelta(days=1).total_seconds()
release_start = datetime(2002, 1, 1, 0, 0)
release_end = release_start + runtime_total
release_time = xr.date_range(release_start, release_end, freq="D").values


# Release N_PARTICLES per day homogeneously distributed within a rectangle
nsqrt = np.sqrt(n_particles_per_day).round().astype(int)
re_lons = np.linspace(24, 25, nsqrt)
re_lats = np.linspace(-36, -35, nsqrt)

# Define 3D (time, lon, lat) release cube
release_xx, release_yy = np.meshgrid(re_lons, re_lats)
release_xx = np.tile(release_xx, (release_time.size, 1, 1))
release_yy = np.tile(release_yy, (release_time.size, 1, 1))
release_time = np.tile(release_time[:, None, None], (1, nsqrt, nsqrt))

# Particle set
pset = ParticleSet(
    fieldset=fieldset,
    pclass=JITParticle,
    lon=release_xx,
    lat=release_yy,
    time=release_time,
)

# Kernels
kernel = pset.Kernel(AdvectionRK4)

# Output file
path_target = os.path.join(dir_target, f"trajs.zarr")
output_file = pset.ParticleFile(
    name=path_target,
    outputdt=outputdt,
    chunks=(chunks_particles, chunks_obs),  # (number of particles, timesteps)
)

# Run simulation
ParcelsRandom.seed(1600)

# release particles
pset.execute(
    kernel,
    dt=runtime_dt,
    runtime=runtime_total,
    output_file=output_file,
)

```

</details>

**Comment by erikvansebille** [comment]
Thanks for providing this excellent minimal working example, @nicrie! It really helped me to hone in on the issue and find the bug. It's now been fixed in #1488 

Note that the `obs` dimension would still be the size of your `chunk_obs` (so padded with `NaN`s if you have less observations than `chunk_obs`), but that is by design. It's a way to drastically speed up output writing, as the array doesn't have to be extended every new observation, but is set at the size of `chunks_obs` initially. See the implementation below

https://github.com/OceanParcels/parcels/blob/3cf8b2ac7fe0b7f4da8a6dc7536b4dff85d8c183/parcels/particlefile.py#L272-L273
and
https://github.com/OceanParcels/parcels/blob/3cf8b2ac7fe0b7f4da8a6dc7536b4dff85d8c183/parcels/particlefile.py#L179-L184

**Comment by nicrie** [comment]
Thank you @erikvansebille for the quick fix! I'll update to the new `master` and will run it with some new simulations. Really enjoying `parcels` so far :)
